### PR TITLE
Feature/t8n

### PR DESCRIPTION
**Pull request by yerke26**
Fixes Closes Resolves #

_Please choose one of the keywords above to refer to the issue this PR solves followed by the issue number (e.g. Fixes #000). If no issue number, remove the line. Also, remove everything marked optional that is not applicable. Remove this note after reading._

## Changes

- _List the changes_

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [x] Yes
- [ ] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [ ] Yes
- [ ] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus.Ethash/EthashDifficultyCalculator.cs [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 [assembly: InternalsVisibleTo("Ethereum.Test.Base")]
 [assembly: InternalsVisibleTo("Ethereum.Difficulty.Test")]
+[assembly: InternalsVisibleTo("Evm")]
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Consensus.Ethash/EthashDifficultyCalculator.cs:13]: why was this added here?

Comment (yerke26) [reply src/Nethermind/Nethermind.Consensus.Ethash/EthashDifficultyCalculator.cs:13]: Used to calculate difficulty in ApplyMergeChecks as per logic of geth

File: src/Nethermind/Nethermind.Core/BaseFeeCalculator.cs [hunk 2]
```diff
@@ -11,32 +11,37 @@ public static class BaseFeeCalculator
     {
         public static UInt256 Calculate(BlockHeader parent, IEip1559Spec specFor1559)
         {
-            UInt256 expectedBaseFee = parent.BaseFeePerGas;
+            return Calculate(parent.BaseFeePerGas, parent.GasUsed, parent.GasLimit, parent.Number, specFor1559);
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Core/BaseFeeCalculator.cs:14]: why was this extra call added ?

File: tools/evm/T8NTool/T8NTool.cs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        string stateFork,
+        string? stateReward)
+    {
+        Dictionary<Address, AccountState> allocJson = _ethereumJsonSerializer.Deserialize<Dictionary<Address, AccountState>>(File.ReadAllText(inputAlloc));
+        EnvInfo envInfo = _ethereumJsonSerializer.Deserialize<EnvInfo>(File.ReadAllText(inputEnv));
+        Transaction[] transactions;
+        if (inputTxs.EndsWith(".json")) {
+            TransactionInfo[] txInfoList = _ethereumJsonSerializer.Deserialize<TransactionInfo[]>(File.ReadAllText(inputTxs));
+            transactions = txInfoList.Select(txInfo => txInfo.ConvertToTx()).ToArray();
+        } else {
```
Comment (Demuirgos) [inline tools/evm/T8NTool/T8NTool.cs:97]: if not json we assume it is an RLP encoded tx List?

File: tools/evm/T8NTool/T8NTool.cs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+            FileInfo fileInfo = new(basedir + filename);
+            Directory.CreateDirectory(fileInfo.DirectoryName!);
+            using StreamWriter writer = new(fileInfo.FullName);
+            writer.Write(_ethereumJsonSerializer.Serialize(outputObject, true));
+        }
+    }
+
+    private bool IsPostMerge(IReleaseSpec spec)
+    {
+        return spec != London.Instance &&
```
Comment (Demuirgos) [inline tools/evm/T8NTool/T8NTool.cs:266]: why not check the other way around? since there are fewer forks after the merge?

File: src/Nethermind/Nethermind.Core/BaseFeeCalculator.cs (hunk 2 above)
Comment (yerke26) [reply src/Nethermind/Nethermind.Core/BaseFeeCalculator.cs]: thought it would be convenient at some point, now reverted

File: tools/evm/T8NTool/T8NTool.cs (hunk 4 above)
Comment (yerke26) [reply tools/evm/T8NTool/T8NTool.cs]: fixed

File: tools/evm/T8NTool/T8NTool.cs [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        Directory.CreateDirectory(fileInfo.DirectoryName!);
+        using StreamWriter writer = new(fileInfo.FullName);
+        writer.Write(_ethereumJsonSerializer.Serialize(outputObject, true));
+    }
+
+    private bool IsPostMerge(IReleaseSpec spec)
+    {
+        return spec == ArrowGlacier.Instance
+               || spec == GrayGlacier.Instance
+               || spec == Paris.Instance
```
Comment (marcindsobczak) [inline tools/evm/T8NTool/T8NTool.cs:280]: Glaciers are pre-merge
```suggestion
        return spec == Paris.Instance
```

File: tools/evm/T8NTool/T8NTool.cs (hunk 3 above)
Comment (Demuirgos) [reply tools/evm/T8NTool/T8NTool.cs:116]: What if a file not ending in either .rlp or .json was provided? We will consider it an rlp file?

File: src/Nethermind/Ethereum.Test.Base/AccountState.cs [hunk 6]
```diff
@@ -21,16 +21,23 @@ public class AccountState
...
             var code = stateProvider.GetCode(address);
-            return new AccountState()
+            var accountState = new AccountState()
             {
                 Nonce = account.Nonce,
                 Balance = account.Balance,
                 Code = code.Length == 0 ? null : code
             };
+
+            if (storages.TryGetValue(address, out var storage))
```
Comment (Demuirgos) [inline src/Nethermind/Ethereum.Test.Base/AccountState.cs:35]: what is the goal of this change?

File: tools/evm/T8NTool/T8NToolTracer.cs [hunk 7]
```diff
@@ -0,0 +1,242 @@
+﻿using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Evm;
+using Nethermind.Evm.Tracing;
+using Nethermind.Int256;
+
+namespace Evm.T8NTool
+{
+    public class T8NToolTracer : IBlockTracer, ITxTracer, IJournal<int>, ITxTracerWrapper
```
Comment (Demuirgos) [inline tools/evm/T8NTool/T8NToolTracer.cs:9]: did you check if GethLikeTxMemoryTracer and GethLikeBlockMemoryTracer aren't suitable for this?

File: src/Nethermind/Ethereum.Test.Base/AccountState.cs (hunk 6 above)
Comment (yerke26) [reply src/Nethermind/Ethereum.Test.Base/AccountState.cs:35]: Alloc output needs storage values, so I created T8NToolTracer in order to collect storages while executing contracts

File: tools/evm/T8NTool/AccountStateConverter.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        if (!accountState.Nonce.IsZero)
+        {
+            writer.WritePropertyName("nonce");
+            _uInt256Converter.Write(writer, accountState.Nonce, options);
+        }
+
+        if (!accountState.Storage.IsNullOrEmpty())
+        {
+            writer.WritePropertyName("storage");
+            var storage = accountState.Storage.ToDictionary(pair => Utils.ConvertToHash256(pair.Key), pair => Utils.ConvertToHash256(pair.Value));
```
Comment (LukaszRozmej) [inline tools/evm/T8NTool/AccountStateConverter.cs:48]: unnecessary allocation, can be done in foreach

Comment (LukaszRozmej) [reply tools/evm/T8NTool/AccountStateConverter.cs:48]: better we can now serialize those dictionaries with referenced PR

File: tools/evm/JsonTypes/TransactionInfo.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+using Nethermind.Core;
+using Nethermind.Core.Eip2930;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Test.Builders;
+using Nethermind.Crypto;
+using Nethermind.Int256;
+
+namespace Evm.JsonTypes
+{
+    public class TransactionInfo
```
Comment (LukaszRozmej) [inline tools/evm/JsonTypes/TransactionInfo.cs:11]: can `TransactionForRpc` be reused?

**Review by LukaszRozmej — COMMENTED** [review]
Didn't review everything yet.
Do we have to push those json files? Aren't they from some other repo that we can just submodule?

File: src/Nethermind/Nethermind.Consensus/Rewards/RewardCalculator.cs [hunk 10]
```diff
@@ -11,17 +11,31 @@ namespace Nethermind.Consensus.Rewards
...
-        private readonly ISpecProvider _specProvider;
+        private readonly ISpecProvider? _specProvider;
+        private readonly UInt256? _blockReward;
 
         public RewardCalculator(ISpecProvider? specProvider)
         {
             _specProvider = specProvider ?? throw new ArgumentNullException(nameof(specProvider));
         }
 
+        public RewardCalculator(UInt256? blockReward)
```
[the remark this answers is not in the document]
Comment (yerke26) [reply src/Nethermind/Nethermind.Consensus/Rewards/RewardCalculator.cs:22]: StaticRewardCalculator just sets the reward for blocks, does not calculate anything, just a converter from IDictionary<long, UInt256> to List<BlockRewardInfo>. I need RewardCalculator because it does calculate reward for uncles and the main reward. I'm passing static blockReward value into constructor because It is specified in the tests (state.reward flag) and IReleaseSpec's reward value is not  the thing that I need (as per tests).

File: tools/evm/JsonTypes/TransactionInfo.cs (hunk 9 above)
Comment (yerke26) [reply tools/evm/JsonTypes/TransactionInfo.cs]: I tried to use TransactionForRpc, but did some modifications, added some fields, but some tests are failing

File: tools/evm/T8NTool/ReceiptJsonConverter.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+using Nethermind.Int256;
+using Nethermind.Serialization.Json;
+
+namespace Evm.T8NTool;
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Nethermind.Core;
+
+public class ReceiptJsonConverter : JsonConverter<TxReceipt>
```
[the remark this answers is not in the document]
Comment (yerke26) [reply tools/evm/T8NTool/ReceiptJsonConverter.cs:15]: No, I tried, but created ReceiptJsonConverter in order to write receipt as json exactly as in geth, ReceiptForRpc differs a lot

File: tools/evm/T8NTool/AccountStateConverter.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        return _ethereumJsonSerializer.Deserialize<AccountState>(ref reader);
+    }
+
+    public override void Write(Utf8JsonWriter writer, AccountState accountState, JsonSerializerOptions options)
+    {
+        writer.WriteStartObject();
+
+        if (accountState.Code != null)
+        {
+            writer.WritePropertyName("code");
```
[the remark this answers is not in the document]
Comment (yerke26) [reply tools/evm/T8NTool/AccountStateConverter.cs]: done

File: tools/evm/T8NTool/GethErrorMappings.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    private static Dictionary<string, string> _mapping = new()
+    {
+        { WrongTransactionNonceError, WrongTransactionNonceGethError },
+        { MissingTxToError, MissingTxToGethError }
+    };
+ 
+    public static string GetErrorMapping(string error, params object[] arguments)
+    {
+        return _mapping.TryGetValue(error, out var newError) ? string.Format(newError, arguments) : error;
+    }
```
[the remark this answers is not in the document]
Comment (yerke26) [reply tools/evm/T8NTool/GethErrorMappings.cs]: done

File: src/Nethermind/Ethereum.Test.Base/GeneralTestBase.cs [hunk 14]
```diff
@@ -147,19 +147,24 @@ protected EthereumTestResult RunTest(GeneralStateTest test, ITxTracer txTracer)
...
+                        stateProvider.Set(new StorageCell(accountState.Key, storageItem.Key),
+                            storageItem.Value.WithoutLeadingZeros().ToArray());
+                    }
                 }
 
-                stateProvider.CreateAccount(accountState.Key, accountState.Value.Balance);
-                stateProvider.InsertCode(accountState.Key, accountState.Value.Code, specProvider.GenesisSpec);
-                stateProvider.SetNonce(accountState.Key, accountState.Value.Nonce);
+                stateProvider.CreateAccount(accountState.Key, accountState.Value.Balance, accountState.Value.Nonce);
+                if (accountState.Value.Code != null)
```
[the remark this answers is not in the document]
Comment (yerke26) [reply src/Nethermind/Ethereum.Test.Base/GeneralTestBase.cs]: done

File: src/Nethermind/Ethereum.Test.Base/BlockchainTestBase.cs [hunk 15]
```diff
@@ -111,19 +111,9 @@ protected async Task<EthereumTestResult> RunTest(BlockchainTest test, Stopwatch?
...
-                               test.Network != ConstantinopleFix.Instance &&
-                               test.Network != Constantinople.Instance &&
-                               test.Network != Byzantium.Instance &&
-                               test.Network != SpuriousDragon.Instance &&
-                               test.Network != TangerineWhistle.Instance &&
-                               test.Network != Dao.Instance &&
-                               test.Network != Homestead.Instance &&
-                               test.Network != Frontier.Instance &&
-                               test.Network != Olympic.Instance;
+            bool isPostMerge = test.Network == Paris.Instance
```
[the remark this answers is not in the document]
Comment (yerke26) [reply src/Nethermind/Ethereum.Test.Base/BlockchainTestBase.cs]: done

File: tools/evm/T8NTool/AccountStateConverter.cs (hunk 8 above)
Comment (yerke26) [reply tools/evm/T8NTool/AccountStateConverter.cs]: removed AccountStateConverter, used NativePrestateTracerAccount and its converter

File: tools/evm/Program.cs [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+            var outputBodyOpt = new Option<string>("--output.body", description: "Output body");
+            var outputResultOpt = new Option<string>("--output.result", description: "Output result");
+            var stateChainIdOpt = new Option<ulong>("--state.chainId", description: "State chain id", getDefaultValue: () => 1);
+            var stateForkOpt = new Option<string>("--state.fork", description: "State fork", getDefaultValue: () => "GrayGlacier");
+            var stateRewardOpt = new Option<string>("--state.reward", description: "State reward");
+            var traceMemoryOpt = new Option<bool>("--trace.memory", description: "Trace memory", getDefaultValue: () => false);
+            var traceOpt = new Option<bool>("--trace", description: "Configures the use of the JSON opcode tracer. This tracer emits traces to files as trace-<txIndex>-<txHash>.jsonl", getDefaultValue: () => false);
+            var traceNoStackOpt = new Option<bool>("--trace.noStack", description: "Trace no stack", getDefaultValue: () => false);
+            var traceReturnDataOpt = new Option<bool>("--trace.returnData", description: "Trace return data", getDefaultValue: () => false);
```
Comment (danceratopz) [inline tools/evm/Program.cs:38]: Would it possible to use strict lower-case for the cli flags? Or make them case-insensitive?

We can accommodate variations, but until now all other tools supported by EEST use lowercase as apposed to camel case (see also `basedir`), which makes it easier to abstract the interface for all tools.

```suggestion
            var inputAllocOpt = new Option<string>("--input.alloc", description: "Input allocations", getDefaultValue: () => "alloc.json");
            var inputEnvOpt = new Option<string>("--input.env", description: "Input environment", getDefaultValue: () => "env.json");
            var inputTxsOpt = new Option<string>("--input.txs", description: "Input transactions", getDefaultValue: () => "txs.json");
            var outputAllocOpt = new Option<string>("--output.alloc", description: "Output allocations");
            var outputBaseDirOpt = new Option<string>("--output.basedir", description: "Output base directory");
            var outputBodyOpt = new Option<string>("--output.body", description: "Output body");
            var outputResultOpt = new Option<string>("--output.result", description: "Output result");
            var stateChainIdOpt = new Option<ulong>("--state.chainid", description: "State chain id", getDefaultValue: () => 1);
            var stateForkOpt = new Option<string>("--state.fork", description: "State fork", getDefaultValue: () => "GrayGlacier");
            var stateRewardOpt = new Option<string>("--state.reward", description: "State reward");
            var traceMemoryOpt = new Option<bool>("--trace.memory", description: "Trace memory", getDefaultValue: () => false);
            var traceOpt = new Option<bool>("--trace", description: "Configures the use of the JSON opcode tracer. This tracer emits traces to files as trace-<txIndex>-<txHash>.jsonl", getDefaultValue: () => false);
            var traceNoStackOpt = new Option<bool>("--trace.nostack", description: "Trace no stack", getDefaultValue: () => false);
            var traceReturnDataOpt = new Option<bool>("--trace.returndata", description: "Trace return data", getDefaultValue: () => false);
```

File: tools/evm/JsonTypes/EnvInfo.cs [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+                throw new T8NException("post-cancun env requires parentBeaconBlockRoot to be set", ExitCodes.ErrorConfig);
+            }
+        }
+
+        private void ApplyMergeChecks(ISpecProvider specProvider)
+        {
+            if (specProvider.TerminalTotalDifficulty?.IsZero ?? false)
+            {
+                if (CurrentRandom == null) throw new T8NException("post-merge requires currentRandom to be defined in env", ExitCodes.ErrorConfig);
+                if (CurrentDifficulty?.IsZero ?? false) throw new T8NException("post-merge difficulty must be zero (or omitted) in env", ExitCodes.ErrorConfig);
```
Comment (danceratopz) [inline tools/evm/JsonTypes/EnvInfo.cs:101]: The following `env.json` (with a `currentlDifficulty` value of `"0"`) and `--state.fork Cancun` unexpectedly triggers this exception.

```json
{
    "currentCoinbase": "0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
    "currentGasLimit": "100000000000000000",
    "currentNumber": "1",
    "currentTimestamp": "1000",
    "currentRandom": "0",
    "currentDifficulty": "0",
    "currentBaseFee": "7",
    "currentExcessBlobGas": "0",
    "currentBlobGasUsed": "0",
    "parentUncleHash": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
    "parentBeaconBlockRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "blockHashes": {},
    "ommers": [],
    "withdrawals": []
}
```

I think this resolves it?
```suggestion
                if (CurrentDifficulty != null && !CurrentDifficulty.Value.IsZero) throw new T8NException("post-merge difficulty must be zero (or omitted) in env", ExitCodes.ErrorConfig);
```

**Review by danceratopz — COMMENTED** [review]
Great to see this PR! :stars: 

I just tried the `t8n` interface out and stumbled over a couple of things, comments below!

I did a short test of isolated commands; I hope to integrate it fully in [ethereum/execution-spec-tests](https://github.com/ethereum/execution-spec-tests) (EEST) very soon, which might trigger more feedback :slightly_smiling_face:.
